src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -77,6 +82,130 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
     networkControllerRef ! SendToNetwork(msg, SendToRandom)
   }
 
+  /**
+    * Helper method which is deciding whether chain is likely nearly or fully synchronized with the network
+    */
+  private def chainAlmostDownloaded: Boolean = {
+    historyReaderOpt.exists { historyReader =>
+      (historyReader.headersHeight - historyReader.fullBlockHeight < 3) &&
+       historyReader.bestHeaderOpt.exists(_.isNew(timeProvider, 1.hour))
+    }
+  }
+
+  /**
+    * Logic to process block parts got from another peer.
+    * Filter out non-requested block parts (with a penalty to spamming peer),
+    * parse block parts and send valid modifiers to NodeViewHolder
+    *
+    * Also, re-announce new blocks.
+    */
+  override protected def modifiersFromRemote(data: ModifiersData, remote: ConnectedPeer): Unit = {
+    val typeId = data.typeId
+    val modifiers = data.modifiers
+    log.info(s"Got ${modifiers.size} modifiers of type $typeId from remote connected peer: $remote")
+    log.trace(s"Received modifier ids ${modifiers.keySet.map(encoder.encodeId).mkString(",")}")
+
+    // filter out non-requested modifiers
+    val requestedModifiers = processSpam(remote, typeId, modifiers)
+
+    Constants.modifierSerializers.get(typeId) match {
+      case Some(serializer: ScorexSerializer[ErgoTransaction]@unchecked) if typeId == Transaction.ModifierTypeId =>
+        // parse all transactions and send them to node view holder
+        val parsed: Iterable[ErgoTransaction] = parseModifiers(requestedModifiers, serializer, remote)
+        viewHolderRef ! TransactionsFromRemote(parsed)
+
+      case Some(serializer: ScorexSerializer[ErgoPersistentModifier]@unchecked) =>
+        // parse all modifiers and put them to modifiers cache
+        val parsed: Iterable[ErgoPersistentModifier] = parseModifiers(requestedModifiers, serializer, remote)
+        val valid = parsed.filter(validateAndSetStatus(remote, _))
+        if (valid.nonEmpty) viewHolderRef ! ModifiersFromRemote[ErgoPersistentModifier](valid)
+
+        // If chain is synced or almost synced, announce new modifiers received.
+        // Helping to push new blocks around the network faster.
+        if (chainAlmostDownloaded) {
+          val toAnnounce = valid.map(_.id).toSeq
+          log.info(s"Announcing recent modifiers: ${toAnnounce.size}, type: $typeId}")
+          val msg = Message(invSpec, Right(InvData(typeId, toAnnounce)), None)
```

Comment:
I wonder, why the source is None?

Reply:
None means there's no external source for the message (so it is sent from the node).